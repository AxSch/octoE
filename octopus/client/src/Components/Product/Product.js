import React, { useState } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus, faMinus } from '@fortawesome/free-solid-svg-icons';
import ProductStyling from './Product.styled'
import { storeItems } from '../../reducers/cart/cartSlice'
import store from '../../store/store'

const {
    StyledProductImg,
    Column,
    StyledProduct,
    StyledProductHeader,
    StyledProductMetaSection,
    StyledProductQuantity,
    StyledProductQuantityCTA,
    StyledProductDesc,
    StyledSpecHeader,
    StyledSpecSection,
    StyledColumnItem,
    StyledQtyRemoveButton,
    StyledQtyAddButton,
    StyledQtyAmount,
    StyledProductPriceSection,
    StyledProductPriceRow,
    StyledPrimaryPrice,
    StyledSecondaryPrice,
    StyledPricing,
    StyledCartButtonRow,
    StyledCartButton,
    StyledProductDescHeader
} = ProductStyling

const Product = ({ data }) => {
    const [quantity, setQuantity] = useState(1)
    
    const formatPrice = (price) => {
        const decimalPoint = (((price / 1000) * 10).toFixed(2)).toString()
        const splitAtDecimal = decimalPoint.split('.')
        return {
            primaryNo: Number(splitAtDecimal[0]),
            secondaryNo: Number(splitAtDecimal[1]),
        }
    }
    const formattedPrice = formatPrice(data.price)
    
    const addToCart = () => {
        const cartItem = {
            id: data.id,
            productName: data.name,
            price: data.price,
            quantity: quantity
        }
        store.dispatch(storeItems(cartItem))
    }

    const reduceQuantity = () => {
        if (quantity === 1) {
            return 1
        }
        setQuantity(quantity - 1)
    }

    return (
        <StyledProduct>
            <StyledProductImg>
                <img alt="product-main" src={data.imgUrl} />
            </StyledProductImg>
            <StyledProductHeader>
                <h1>{data.name}</h1>
            </StyledProductHeader>
            <StyledProductMetaSection>{data.power} &#47;&#47; Packet of {data.quantity}</StyledProductMetaSection>
            <StyledProductPriceSection>
                <StyledProductPriceRow>
                    <StyledPricing>
                        <StyledPrimaryPrice>
                            <span>{formattedPrice.primaryNo}</span>
                        </StyledPrimaryPrice>
                        <StyledSecondaryPrice>
                            <span>.{formattedPrice.secondaryNo}</span>
                        </StyledSecondaryPrice>
                    </StyledPricing>
                    <StyledProductQuantityCTA>
                        <StyledProductQuantity>
                            QTY
                        </StyledProductQuantity>
                        <StyledQtyRemoveButton onClick={() => reduceQuantity()}>
                            <FontAwesomeIcon icon={faMinus} />
                        </StyledQtyRemoveButton>
                        <StyledQtyAmount>{quantity}</StyledQtyAmount>
                        <StyledQtyAddButton onClick={() => setQuantity(quantity + 1)}>
                            <FontAwesomeIcon icon={faPlus} />
                        </StyledQtyAddButton>
                    </StyledProductQuantityCTA>
                </StyledProductPriceRow>
                <StyledCartButtonRow>
                    <StyledCartButton onClick={() => addToCart()}>Add to cart</StyledCartButton>
                </StyledCartButtonRow>
            </StyledProductPriceSection>
            <StyledProductDesc>
                <StyledProductDescHeader>
                    <h1>Description</h1>
                </StyledProductDescHeader>
                <Column>
                    <span>{data.description}</span>
                </Column>
            </StyledProductDesc>
            <StyledSpecSection>
                <StyledSpecHeader>
                    <h1>Specifications</h1>
                </StyledSpecHeader>
                <Column>
                    <StyledColumnItem>
                        <Column>
                            <span>Brand</span>
                        </Column>
                        <Column>{data.brand}</Column>
                    </StyledColumnItem>
                    <StyledColumnItem>
                        <Column>
                            <span>Item Weight</span>
                        </Column>
                        <Column>{data.weight}</Column>
                    </StyledColumnItem>
                    <StyledColumnItem>
                        <Column>
                            <span>Dimensions</span>
                        </Column>
                        <Column>{data.height}x{data.width}x{data.length}</Column>
                    </StyledColumnItem>
                    <StyledColumnItem>
                        <Column>
                            <span>Item model number</span>
                        </Column>
                        <Column>{data.modelCode}</Column>
                    </StyledColumnItem>
                    <StyledColumnItem>
                        <Column>
                            <span>Colour</span>
                        </Column>
                        <Column>{data.colour}</Column>
                    </StyledColumnItem>
                </Column>
            </StyledSpecSection>
        </StyledProduct>
    )
}

export default Product;
