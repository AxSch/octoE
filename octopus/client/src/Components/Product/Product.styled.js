import styled from 'styled-components'


const StyledProductImg = styled.div`
    display: flex;
    justify-content: center;
    > img {
        width: 330px;
        height: 390px;
    }
`

const StyledProduct = styled.div`
    background-image: linear-gradient(to bottom, rgb(5, 15, 42) 100%, rgb(1,25,59) 100%);
    @media(min-width:768px) {
        padding-top: 40px;
        background-color: rgb(1,25,59);
    }
`
const StyledProductHeader = styled.div`
    display: flex;
    flex-direction:row;
    background-color: rgb(1,25,59);
    color: white;
    font-size: 18px;
    padding: 0 30px;
    @media (min-width:640px) {
        justify-content: center;
    }
    
`

const StyledProductMetaSection = styled.div`
    background-color: rgb(1,25,59);
    color: rgb(54,89,141);
    font-size: 16px;
    padding: 0 0 20px 30px;
    @media (min-width:640px) {
        display: flex;
        justify-content: center;
    }
`

const StyledProductPriceSection = styled.div`
    background-color: rgb(5, 15, 42);
    color: white;
    font-size: 16px;
    padding: 20px 30px 30px 30px;
    display: flex;
    flex-wrap: wrap;
`

const Column = styled.div`
    display: flex;
    flex-direction: column;
    width: 100%;
    @media (min-width:1024px) {
        width: 60%;
        align-items: space-around;
    }  
`

const StyledProductPriceRow = styled.div`
    display: flex;
    padding-bottom: 18px;
    width: 100%;
    align-items: center;
    @media(min-width:640px) {
        align-items: end;
    }
`

const StyledCartButton = styled.button`
    color: white;
    font-size: 26px;
    font-weight: 700;
    border: none;
    border-radius: 6px;
    background-image: linear-gradient(to bottom, rgb(255,100,170) 50%, rgb(251, 26, 123) 50%);
    width: 100%;
    padding: 16px;
    cursor: pointer;
    &:hover {
        color: rgb(242, 242, 242);
        background-image: linear-gradient(to bottom, rgb(255, 153, 199) 50%, rgb(253, 104, 169) 50%);
    }

    @media (min-width:640px) {
        width: 70%;
        margin-top: 12px;
    }
    @media (min-width:768px) {
        width: 30%;
        margin-top: 12px;
    }
`

const StyledProductQuantity = styled.div`
    display: flex;
    width: 100%;
    justify-content: end;
    padding-right: 26px;
    font-size: 8px;
    color: rgb(54,89,141);
    @media (min-width:640px) {
        align-items: center;
        justify-content: start;
        font-size: 16px;
        margin-left: 34px;
    }

`

const StyledProductQuantityCTA = styled.div`
    display: flex;
    width: 100%;
    justify-content: end;
    flex-wrap: wrap;
    @media (min-width:640px) {
        align-items: center;
        justify-content: start;
        margin-left: 26px;
    }
`

const StyledProductDesc = styled.div`
    display: flex;
    background-color: rgb(1,25,59);
    color: white;
    flex-wrap: wrap;
    padding: 10px 30px 20px 30px; 
    @media (min-width:640px) {
        justify-content: center;
        align-items: center;
    }
`

const StyledProductDescHeader = styled.div`
@media (min-width:1024px) {
    margin-right: 40px;
    padding-right: 56px; 
}
`

const StyledSpecHeader = styled.div`
    display: flex;
    padding-bottom: 10px;
    @media (min-width:1024px) {
        margin-right: 40px; 
    }
`

const StyledSpecSection = styled.div`
    color: white;
    padding: 30px 30px 30px 26px; 
    background-color: rgb(5, 15, 42);
    display: flex;
    flex-wrap: wrap;
    @media (min-width:640px) {
        justify-content: center;
    } 
`

const StyledColumnItem = styled.div`
    padding: 20px 0;
    border-bottom: 2px solid rgb(1,25,59);
    display: flex;
`

const StyledQtySubtractButton = styled.button`
    background-color: rgb(26, 42, 67);
    color: white;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    &:hover {
        background-color: rgb(36, 57, 92);
    }
    @media (min-width:640px) {
        font-size: 22px;
        padding: 10px;
    }
`

const StyledQtyAddButton = styled.button`
    background-color: rgb(54, 89, 141);
    color: white;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    &:hover {
        background-color: rgb(63, 104, 166);
    }
    @media (min-width:640px) {
        font-size: 22px;
        padding: 10px;
    }
`

const StyledQtyAmount = styled.div`
    font-weight: 600;
    padding: 0 6px;
    display: flex;
    align-items: center;
`

const StyledPrimaryPrice = styled.div`
    display: flex;
    flex-direction: column;
    font-weight: 600;
    font-size: 22px;
    @media (min-width:640px) {
        font-size: 32px;
    }
`

const StyledSecondaryPrice = styled.div`
    display: flex;
    flex-direction: column;
    font-weight: 600;
    font-size: 14px;
    @media (min-width:640px) {
        font-size: 16px;
    }
`

const StyledPricing = styled.div`
    display: flex;
    @media (min-width:640px) {
        width:100%;
        justify-content: end;
    }   
`

const StyledCartButtonRow = styled.div`
    display: flex;
    width:100%;
    @media (min-width:640px) {
        justify-content: center;
    }   
`

export default {
    StyledProductImg,
    Column,
    StyledProduct,
    StyledProductHeader,
    StyledProductMetaSection,
    StyledProductQuantity,
    StyledProductQuantityCTA,
    StyledProductDesc,
    StyledSpecHeader,
    StyledSpecSection,
    StyledColumnItem,
    StyledQtySubtractButton,
    StyledQtyAddButton,
    StyledQtyAmount,
    StyledProductPriceSection,
    StyledProductPriceRow,
    StyledPrimaryPrice,
    StyledSecondaryPrice,
    StyledPricing,
    StyledCartButtonRow,
    StyledCartButton,
    StyledProductDescHeader
}