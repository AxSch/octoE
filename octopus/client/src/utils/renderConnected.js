import React from 'react';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import cartSlice from '../reducers/cart/cartSlice';

const renderConnected = (
  ui, {
    store = createStore(cartSlice, {}),
    ...renderOptions
  } = {}
) => {
  const Wrapper = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );
  return render(ui, { wrapper: Wrapper, ...renderOptions});
};

export default renderConnected;
