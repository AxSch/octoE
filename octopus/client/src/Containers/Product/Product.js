import React from 'react'
import { GET_PRODUCT } from '../../api/productGQL'
import { useQuery } from "@apollo/react-hooks";
import { Product } from '../../Components/Product/Product';


export const ProductContainer = () => {
    const { loading, error, data } = useQuery(GET_PRODUCT, {
        variables: { ID: 1 }
    })
    if (loading) return <div>Loading...</div>;
    if (error) return <div>Error!</div>;
    return(
        <Product data={data.product} />
    )
}
