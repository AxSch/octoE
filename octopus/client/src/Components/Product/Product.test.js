import { fireEvent } from '@testing-library/react';
import * as ReactRedux from 'react-redux';
import React from 'react'
import Product from './Product'

import renderConnected from '../../utils/renderConnected'

const initialState = {
  items: {}
}

const mockProduct = {
  id: 1,
  name: "Bulba",
  power: "Green",
  description: "Renewables at your fingertips",
  price: 3444,
  quantity: 1,
  brand: "Oak",
  weight: 15.2,
  height: 2.04,
  length: 1.00
}
let getByTestId, container

describe('should be able to increase and decrease product quantity', () => {

  beforeEach(() => {
    const utils = renderConnected(<Product data={mockProduct}/>, { initialState });
    getByTestId = utils.getByTestId;
    container = utils.container
  });

  it('renders successfully', () => {
    expect(getByTestId('product')).toBeInTheDocument()
    expect(container).toMatchSnapshot();
  });
  
  it('should increase the quantity', () => {
    const addButton = getByTestId('addButton')
    const quantityCounter = getByTestId('productQuantity')
    expect(quantityCounter.textContent).toBe('1')
    fireEvent.click(addButton)
    expect(quantityCounter.textContent).toBe('2')
  });

  it('should decrease the quantity', () => {
    const addButton = getByTestId('addButton')
    const subtractButton = getByTestId('subtractButton')
    const quantityCounter = getByTestId('productQuantity')
    expect(quantityCounter.textContent).toBe('1')
    fireEvent.click(addButton)
    expect(quantityCounter.textContent).toBe('2')
    fireEvent.click(subtractButton)
    expect(quantityCounter.textContent).toBe('1')
  });
});

describe('should be able to add items to the basket', () => {
  const mockDispatch = jest.fn()
  ReactRedux.useDispatch = jest.fn().mockImplementation(() => mockDispatch);
  
  beforeEach(() => {
    const utils = renderConnected(<Product data={mockProduct}/>, { initialState });
    getByTestId = utils.getByTestId;
  });
  
  it('should successfully add an item to the basket', () => {
    const addToCartButton = getByTestId('addToCartButton')
    fireEvent.click(addToCartButton)
    const expectedPayload = {
      payload: {
        id: 1,
        price: 3444,
        productName: "Bulba",
        quantity: 1,
      },
      type: "cart/storeItems"
    }
    expect(mockDispatch).toHaveBeenCalledWith(expectedPayload)
  });
});
