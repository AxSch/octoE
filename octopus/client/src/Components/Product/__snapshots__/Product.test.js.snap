// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to increase and decrease product quantity renders successfully 1`] = `
<div>
  <div
    class="sc-gsTCUz ehKVQH"
    data-testid="product"
  >
    <div
      class="sc-bdfBwQ hTgnrq"
    >
      <img
        alt="product-main"
      />
    </div>
    <div
      class="sc-dlfnbm dfaYQX"
    >
      <h1>
        Bulba
      </h1>
    </div>
    <div
      class="sc-hKgILt egFqBQ"
    >
      Green
       // Packet of 
      1
    </div>
    <div
      class="sc-eCssSg gLBkuG"
    >
      <div
        class="sc-gKsewC ioURlo"
      >
        <div
          class="sc-bkzZxe kLiOVN"
        >
          <div
            class="sc-fodVxV JsNvX"
          >
            <span>
              34
            </span>
          </div>
          <div
            class="sc-fFubgz kGOoPf"
          >
            <span>
              .
              44
            </span>
          </div>
        </div>
        <div
          class="sc-pFZIQ gqKuOQ"
        >
          <div
            class="sc-fubCfw ceTAHb"
          >
            QTY
          </div>
          <button
            class="sc-bqyKva chPCpe"
            data-testid="subtractButton"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-minus fa-w-14 "
              data-icon="minus"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="sc-hBEYos jQBlmP"
            data-testid="productQuantity"
          >
            1
          </div>
          <button
            class="sc-kstrdz dUxZPy"
            data-testid="addButton"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-plus fa-w-14 "
              data-icon="plus"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="sc-idOhPF fGMGEh"
      >
        <button
          class="sc-iBPRYJ buoooV"
          data-testid="addToCartButton"
        >
          Add to cart
        </button>
      </div>
    </div>
    <div
      class="sc-jrAGrp dmMqcy"
    >
      <div
        class="sc-kEjbxe dvQzaG"
      >
        <h1>
          Description
        </h1>
      </div>
      <div
        class="sc-jSgupP EWObi"
      >
        <span>
          Renewables at your fingertips
        </span>
      </div>
    </div>
    <div
      class="sc-crrsfI cNFJOC"
    >
      <div
        class="sc-iqHYGH jwRLTY"
      >
        <h1>
          Specifications
        </h1>
      </div>
      <div
        class="sc-jSgupP EWObi"
      >
        <div
          class="sc-dQppl jQFIRP"
        >
          <div
            class="sc-jSgupP EWObi"
          >
            <span>
              Brand
            </span>
          </div>
          <div
            class="sc-jSgupP EWObi"
          >
            Oak
          </div>
        </div>
        <div
          class="sc-dQppl jQFIRP"
        >
          <div
            class="sc-jSgupP EWObi"
          >
            <span>
              Item Weight
            </span>
          </div>
          <div
            class="sc-jSgupP EWObi"
          >
            15.2
          </div>
        </div>
        <div
          class="sc-dQppl jQFIRP"
        >
          <div
            class="sc-jSgupP EWObi"
          >
            <span>
              Dimensions
            </span>
          </div>
          <div
            class="sc-jSgupP EWObi"
          >
            2.04
            x
            x
            1
          </div>
        </div>
        <div
          class="sc-dQppl jQFIRP"
        >
          <div
            class="sc-jSgupP EWObi"
          >
            <span>
              Item model number
            </span>
          </div>
          <div
            class="sc-jSgupP EWObi"
          />
        </div>
        <div
          class="sc-dQppl jQFIRP"
        >
          <div
            class="sc-jSgupP EWObi"
          >
            <span>
              Colour
            </span>
          </div>
          <div
            class="sc-jSgupP EWObi"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
